// angular-pipes/esm2015/math/bytes.pipe.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Pipe } from '@angular/core';
import { isNumberFinite, isPositive, isInteger, toDecimal } from '../utils/utils';
export class BytesPipe {
    /**
     * @param {?} input
     * @param {?=} decimal
     * @param {?=} from
     * @param {?=} to
     * @return {?}
     */
    transform(input, decimal = 0, from = 'B', to) {
        if (!(isNumberFinite(input) &&
            isNumberFinite(decimal) &&
            isInteger(decimal) &&
            isPositive(decimal))) {
            return input;
        }
        /** @type {?} */
        let bytes = input;
        /** @type {?} */
        let unit = from;
        while (unit !== 'B') {
            bytes *= 1024;
            unit = (/** @type {?} */ (BytesPipe.formats[unit].prev));
        }
        if (to) {
            /** @type {?} */
            const format = BytesPipe.formats[to];
            /** @type {?} */
            const result = toDecimal(BytesPipe.calculateResult(format, bytes), decimal);
            return BytesPipe.formatResult(result, to);
        }
        for (const key in BytesPipe.formats) {
            /** @type {?} */
            const format = BytesPipe.formats[key];
            if (bytes < format.max) {
                /** @type {?} */
                const result = toDecimal(BytesPipe.calculateResult(format, bytes), decimal);
                return BytesPipe.formatResult(result, key);
            }
        }
    }
    /**
     * @param {?} result
     * @param {?} unit
     * @return {?}
     */
    static formatResult(result, unit) {
        return `${result} ${unit}`;
    }
    /**
     * @param {?} format
     * @param {?} bytes
     * @return {?}
     */
    static calculateResult(format, bytes) {
        /** @type {?} */
        const prev = format.prev ? BytesPipe.formats[format.prev] : undefined;
        return prev ? bytes / prev.max : bytes;
    }
}
BytesPipe.formats = {
    'B': { max: 1024 },
    'kB': { max: Math.pow(1024, 2), prev: 'B' },
    'KB': { max: Math.pow(1024, 2), prev: 'B' },
    // Backward compatible
    'MB': { max: Math.pow(1024, 3), prev: 'kB' },
    'GB': { max: Math.pow(1024, 4), prev: 'MB' },
    'TB': { max: Number.MAX_SAFE_INTEGER, prev: 'GB' }
};
BytesPipe.decorators = [
    { type: Pipe, args: [{
                name: 'bytes'
            },] }
];
if (false) {
    /** @type {?} */
    BytesPipe.formats;
}
